src/quantizers.jl

```diff
@@ -0,0 +1,153 @@
+    for i = 1:8
+        if (root.children[i].data[1] == r[1] * g[1] * b[1])
+            root.children[i].data[2] += 1
+            ind = i
+            break
+        end
+    end
+    curr = root.children[ind]
+    
+    for i = 2:8
+        cases = map(p->[bitstring(UInt8(p))[6:end], 0, Vector{Int}([]), RGB{N0f8}.(0.0,0.0,0.0), i], 1:8)
+        if (isleaf(curr) == true && i < 8) split!(curr, cases) end
+        if (i == 8)
+            if (isleaf(curr) == true) split!(curr, cases) end
+            for j = 1:8
+                if (curr.children[j].data[1] ==  r[i] * g[i] * b[i])
+                    curr = curr.children[j]
+                    curr.data[2] += 1
+                    push!(curr.data[3], in[1])
+                    curr.data[4] = in[2]
+                    return
+                end
+            end
+        end
+
+        # handle 1:7 cases for rgb to handle first seven bits
+        for j = 1:8
+            if (curr.children[j].data[1] ==  r[i] * g[i] * b[i])
+                curr.children[j].data[2] += 1
+                curr = curr.children[j]
+                break
+            end
+        end
+    end
+end
+
+
+function octreequantisation!(img; numcolors = 256, return_palette = false )
+    # step 1: creating the octree
+    root = Cell(SVector(0.0, 0.0, 0.0), SVector(1.0, 1.0, 1.0), ["root", 0, [], RGB{N0f8}.(0.0, 0.0, 0.0), 0])
+    cases = map(p->[bitstring(UInt8(p))[6:end], 0, Vector{Int}([]), RGB{N0f8}.(0.0,0.0,0.0), 1], 1:8)
+    split!(root, cases)
+    for i in enumerate(img)
+        root.data[2]+=1
+        putin(root, i)
+    end
+    
+    # step 2: reducing tree to a certain number of colors
+    # there is scope for improvements in allleaves as it's found again n again
+    leafs = [p for p  in allleaves(root)]
+    filter!(p -> !iszero(p.data[2]), leafs)
+    tobe_reduced = leafs[1]
+
+    while (length(leafs) > numcolors)
+        parents = unique([parent(p) for p in leafs])
+        parents = sort(parents; by = c -> c.data[2])
+        tobe_reduced = parents[1]
+
+        for i = 1:8
+            append!(tobe_reduced.data[3], tobe_reduced.children[i].data[3])
+            tobe_reduced.data[4] += tobe_reduced.children[i].data[4] * tobe_reduced.children[i].data[2]
+        end
+        tobe_reduced.data[4] /= tobe_reduced.data[2]
+        tobe_reduced.children = nothing
+
+        # we don't want to do this again n again
+        leafs = [p for p  in allleaves(root)]
+        filter!(p -> !iszero(p.data[2]), leafs)
+    end
+
+    # step 3: palette formation  and quantisation now
+    da = [p.data for p in leafs]
+    for i in da
+        for j in i[3]
+            img[j] = i[4]
+        end
+    end
+
+    if return_palette == true
+        colors = [p[4] for p in da]
+        return colors
+    end
+end
+
+function octreequantisation(img; kwargs...)
+    img1 = deepcopy(img)
+    # possible error likely here to return palette kwarg
+    octreequantisation!(img1; kwargs...)
+    return img1
+end
```

Comment:
This seems perfectly a case that can be implemented in a separate standalone package so that other packages (e.g., DitherPunk) can benefit from it.

cc: @adrhill

Reply:
Yes, this would be very useful and I'd be very interested in contributing to (or writing) such a package. :)

Reply:
A small discussion on slack: because this package along with the other two https://github.com/ashwani-rathee/ExifViewer.jl and https://github.com/ashwani-rathee/JPEG2000.jl are part of GSoC 22', we'll quickly get this PR merged and then task switch to JPEG2000. So these quantization methods will be temporarily kept private to GIFImages.